plugins/nf-prov/src/main/nextflow/prov/BcoRenderer.groovy

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright 2022, Seqera Labs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package nextflow.prov
+
+import java.nio.file.Path
+import java.time.format.DateTimeFormatter
+
+import groovy.json.JsonOutput
+import groovy.transform.CompileStatic
+import nextflow.Session
+import nextflow.exception.AbortOperationException
+import nextflow.processor.TaskRun
+import nextflow.script.WorkflowMetadata
+import nextflow.script.params.FileOutParam
+import nextflow.util.CacheHelper
+
+/**
+ * Renderer for the BioCompute Object (BCO) format.
+ *
+ * @author Ben Sherman <bentshermann@gmail.com>
+ */
+@CompileStatic
+class BcoRenderer implements Renderer {
+
+    private WorkflowMetadata metadata
+
+    /**
+     * Get the list of output files for a task.
+     *
+     * @param task
+     */
+    private static List<Path> getTaskOutputs(TaskRun task) {
+        return task
+            .getOutputsByType(FileOutParam)
+            .values()
+            .flatten() as List<Path>
+    }
+
+    /**
+     * Get a mapping of output file to the task that produced it.
+     *
+     * @param tasks
+     */
+    private static Map<Path,TaskRun> getTaskLookup(Set<TaskRun> tasks) {
+        final taskLookup = [:] as Map<Path,TaskRun>
+
+        for( def task : tasks )
+            for( def output : getTaskOutputs(task) )
+                taskLookup[output] = task
+
+        return taskLookup
+    }
+
+    /**
+     * Get the list of workflow inputs. A workflow input is an input file
+     * to a task that was not produced by another task.
+     *
+     * @param tasks
+     * @param taskLookup
+     */
+    private static Set<Path> getWorkflowInputs(Set<TaskRun> tasks, Map<Path,TaskRun> taskLookup) {
+        final workflowInputs = [] as Set<Path>
+
+        tasks.each { task ->
+            task.getInputFilesMap().each { name, path ->
+                if( taskLookup[path] )
+                    return
+
+                workflowInputs << path
+            }
+        }
+
+        return workflowInputs
+    }
+
+    /**
+     * Normalize local paths to remove environment-specific directories.
+     *
+     * @param path
+     */
+    private String normalizePath(Path path) {
+        // TODO: append raw github URL to git assets
+        path.toUriString()
+            .replace(metadata.workDir.toUriString(), '/work')
+            .replace(metadata.projectDir.toUriString(), '')
+            .replace(metadata.launchDir.toUriString(), '')
```

Comment:
Perhpas it would be more "correct" to also replace any leading `/` here, to indicate that the paths are relative rather than absolute?

Reply:
On the other hand, these work paths are currently the only way to determine the task dependencies from the BCO manifest. So maybe we do need to keep them around...

Reply:
Alright! In any case, if they are kept, I think paths better not start with slashes to comply with POSIX conventions (E.g. found [this link in the def](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_324)).